### DISCUSSION TITLE
On the question of primitive atom and derived atoms

### DISCUSSION THREAD

**Discussion by liby** [pmndrs/jotai#692] [Q&A] [answered]
Here's a [simple example](https://codesandbox.io/s/primitive-atom-vs-derived-atoms-u7o1x).

The focus is on the code in the `helper` and `add-item` file sections.

```js
// primitive atom
export const activeAtom = atom({});

// write-only atom, In fact, it can be defined as just atoms 
// but for the sake of observation and comparison, 
// it is deliberately written as read-write atom
export const activeDerivedAtom = atom(
  (get) => {
    const list = get(listAtom);
    if (list.length > 0) {
      return list[list.length - 1];
    }
    return activeAtom;
  },
  (get, set, _update) => {
    const list = get(listAtom);
    set(activeAtom, list[list.length - 1]);
  }
);
```

After defining the code above, I observe that their behavior is very different, and the primitive atom `set` function doesn't work at all.
```js
  const onClick = async () => {
    await setTimeout(() => {
      setList({
        name: `New Item ${formatSecond(new Date())}`,
        id: Date.parse(new Date().toString())
      });

      // Not working
      setActive(list[list.length - 1]);
      // It works
      setActiveDerived();
      console.log("comparison", active, activeDerived);
    }, 1000);
  };
```
![image](https://user-images.githubusercontent.com/38807139/132354652-a98b2a3b-110e-4f49-9350-69c83e28a2b0.png)

This leads me to be very confused, want to find the answer to the question here in you. Maybe there is an obvious answer to the question I asked, but I'm sorry I didn't get it.

@dai-shi

**Comment by dai-shi** [comment, marked as answer]
`return get(activeAtom)` in helper.ts:L31??

**Reply by liby** [reply]
It should be `return {}`. I think I forgot to change it back during Debug, but the change here shouldn't affect the question I'm asking.


https://user-images.githubusercontent.com/38807139/132363373-2696ea98-2edf-4fa4-89c8-85a1dd7d2e70.mp4

**Reply by dai-shi** [reply]
Thanks for the video. Now, I understand what you are seeing.
It took me a while to get it, but I got it. It's not behavior caused by jotai library, but more like JS behavior.
So, `setActive(list[list.length - 1]);` is reading `list` from a stale closure. The `list` is still empty.
This is the behavior you can observe with `useState` too.

**Reply by liby** [reply]
Thank you! It dawned on me that the latest state is not yet available in the current function context.
Also, I would like to ask, in a situation like mine, is it possible to use only primitive atom without using derived atoms? How would I do it?

**Reply by dai-shi** [reply]
```js
const Component = () => {
  const [list, setList] = useState([])
  const [lastItem, setLastItem] = useState()
  const handleClick = () => {
    setTimeout(() => {
      const newItem = 123
      setList(prev => [...prev, newItem])
      setLastItem(list[list.length - 1])
    }, 1000)
  }
  // ...
}
```

This is the useState version.
Because we should not rely that the component will re-render in 1s,
the good solution is `setLastItem(newItem)`.

**Reply by liby** [reply]
`setLastItem(newItem)` also does not solve the issue of the component will re-render in 1s, it is the same as `setActive(list[list.length - 1]);` is reading list from a stale closure. The `list` is still empty.

I think the code to solve this problem should look like this:

```js
const [list, setList] = useState([]);
const lastItem = useMemo(() => {
  return list[list.length - 1];
}, [list]);

const handleClick = () => {
  setTimeout(() => {
    const newItem = {
      name: `New Item ${formatSecond(new Date())}`,
      id: Date.parse(new Date().toString())
    };
    setList((prev) => [...prev, newItem]);
  }, 1000);
};
```

Thank you so much. Today is another day of great benefit. I would like to review these knowledge points, have a nice day!

**Reply by dai-shi** [reply]
Yeah, that's the better solution is this case.
Note: I mean `newItem` must be created without `list`.
